sdc/functions/numpy_like.py

```diff
@@ -0,0 +1,112 @@
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+#     Redistributions of source code must retain the above copyright notice,
+#     this list of conditions and the following disclaimer.
+#
+#     Redistributions in binary form must reproduce the above copyright notice,
+#     this list of conditions and the following disclaimer in the documentation
+#     and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+# THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+# OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+# WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+# OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+# EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+# *****************************************************************************
+
+"""
+
+| This file contains SDC numpy modified functions, rewritten according to concurrency
+
+"""
+
+import numba
+import numpy
+
+from numba.errors import TypingError
+from numba import types
+from numba import jit
+from numba import prange
+from numba import numpy_support
+
+import sdc
+from sdc.datatypes.common_functions import TypeChecker
+from sdc.str_arr_ext import (StringArrayType, pre_alloc_string_array)
+from sdc.utils import sdc_overload
+
+
+def astype(self, dtype):
+    pass
+
+
+@sdc_overload(astype)
+def sdc_astype_overload(self, dtype):
+    """
+    Intel Scalable Dataframe Compiler Developer Guide
+    *************************************************
+    Parallel analogue of numpy 'astype' implementation.
+
+    .. only:: developer
+
+       Test: python -m sdc.runtests sdc.tests.test_sdc_numpy -k astype
+
+    """
+
+    if not isinstance(self, types.Array):
+        return None
+
+    if not isinstance(dtype, (types.StringLiteral, types.UnicodeType, types.Function, types.functions.NumberClass)):
+        ty_checker.raise_exc(dtype, 'string or type', 'dtype')
+
+    if ((isinstance(dtype, types.Function) and dtype.typing_key == str) or
+        (isinstance(dtype, types.StringLiteral) and dtype.literal_value == 'str')):
+        def sdc_astype_number_to_string_impl(self, dtype):
+            num_chars = 0
+            arr_len = len(self)
+
+            # Get total chars for new array
+            for i in prange(arr_len):
+                item = self[i]
+                num_chars += len(str(item))  # TODO: check NA
```

Comment:
It appears that for unicode strings len returns number of chars, but not number of bytes, and allocating StringArray we need to know total number of bytes, not chars in it. Not sure if that's applicable for this case, but still it might be better to use get_utf8_size everywhere, just for consistency.